### PR TITLE
feat(compiler-vapor): slot outlet

### PR DESCRIPTION
**Pull request by LittleSound**
Related: #154 
Related runtime PR: #170

### CODE REVIEW

File: packages/compiler-vapor/src/transforms/transformSlotOutlet.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  let name: SimpleExpressionNode | undefined
+  const nonNameProps: (AttributeNode | DirectiveNode)[] = []
+  const customDirectives: DirectiveNode[] = []
+  for (const p of props) {
+    if (p.type === NodeTypes.ATTRIBUTE) {
+      if (p.value) {
+        if (p.name === 'name') {
+          name = createSimpleExpression(p.value.content, true, p.loc)
+        } else {
+          p.name = camelize(p.name)
```
Comment (sxzz) [inline packages/compiler-vapor/src/transforms/transformSlotOutlet.ts:47]: AST should not be mutated. It'll break cache.

**Review by sxzz — CHANGES_REQUESTED** [review]

**Review by sxzz — CHANGES_REQUESTED** [review]
https://deploy-preview-182--vapor-template-explorer.netlify.app/#eyJzcmMiOiI8c2xvdCB2LXNob3c9XCJva1wiIC8+Iiwib3B0aW9ucyI6e319

Should throw an error when using directives that require runtime (`v-show`...)

File: packages/compiler-vapor/src/transforms/transformSlotOutlet.ts (hunk 1 above)
Comment (LittleSound) [reply packages/compiler-vapor/src/transforms/transformSlotOutlet.ts]: Maybe [this code](https://github.com/vuejs/core-vapor/blob/d3a1df222a0f1009ab26b97c8670a39a36f559db/packages/compiler-vapor/src/transforms/vBind.ts#L61) has a similar problem, do we need to fix it together too?

I may lack knowledge about this caching mechanism, can you briefly describe it? Or how about I can learn about it? Many thanks.

Comment (sxzz) [reply packages/compiler-vapor/src/transforms/transformSlotOutlet.ts]: Mentioned before https://github.com/vuejs/core-vapor/pull/98#discussion_r1468877535, see https://github.com/vuejs/core-vapor/commit/d3baff92b1fb033a1411223a009a8bf68c432037


Cause AST will be reused at the second usage, and could be transformed twice.
https://github.com/vuejs/core-vapor/blob/2d684293f2bd0facf882b1aed126ada5553a2660/packages/compiler-sfc/src/compileTemplate.ts#L226

Comment (sxzz) [reply packages/compiler-vapor/src/transforms/transformSlotOutlet.ts]: The case you mentioned should be fixed.

**Comment by LittleSound** [comment]
When using v-show, there will be no to throw error, this behavior is consistent with compiler-core.
Do we need to fix this situation separately in vapor?

**Comment by sxzz** [comment]
There's a error in non-Vapor mode

https://template-explorer.vuejs.org/#eyJzcmMiOiI8c2xvdCB2LXNob3c9XCJva1wiIC8+Iiwib3B0aW9ucyI6e319

image

**Comment by LittleSound** [comment]
You are right, I was sloppy 🤪. fixed.

File: packages/compiler-vapor/src/transforms/transformSlotOutlet.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+      extend({}, node, { props: nonNameProps }),
+      context as TransformContext<ElementNode>,
+      true,
+    )
+    irProps = isDynamic ? props : [props]
+
+    const { operation } = context.block
+    const directives = operation.filter(
+      oper => oper.type === IRNodeTypes.WITH_DIRECTIVE,
+    ) as WithDirectiveIRNode[]
```
Comment (sxzz) [inline packages/compiler-vapor/src/transforms/transformSlotOutlet.ts:86]:
```suggestion
    const hasDirectives = operation.some(
      oper => oper.type === IRNodeTypes.WITH_DIRECTIVE,
    )
```

File: packages/compiler-vapor/src/generators/slotOutlet.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    ? genExpression(name, context)
+    : ['() => (', ...genExpression(name, context), ')']
+
+  let fallbackArg: CodeFragment[] | undefined
+  if (fallback) {
+    fallbackArg = genBlock(fallback, context)
+  }
+
+  push(NEWLINE, `const n${id} = `)
+  push(
```
Comment (sxzz) [inline packages/compiler-vapor/src/generators/slotOutlet.ts]: These two statement can be merged

File: packages/compiler-vapor/src/transforms/transformSlotOutlet.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        // v-bind shorthand syntax
+        name = createSimpleExpression(
+          camelize(prop.arg.content),
+          false,
+          prop.arg.loc,
+        )
+        name.ast = null
+      }
+    } else {
+      const nonProp = extend({}, prop)
```
Comment (sxzz) [inline packages/compiler-vapor/src/transforms/transformSlotOutlet.ts:63]: Is this necessary?

**Review by sxzz — APPROVED** [review]

**Pull request merged**